architecture/2024-10-deploymenttemplate-controller.md

```diff
@@ -0,0 +1,421 @@
+# Radius DeploymentTemplate Controller
+
+* **Author**: Will Smith (@willdavsmith)
+
+## Overview
+
+Today, users of Radius and future adopters of Radius use many Kubernetes-specific tools in their production workflows. Most of these tools operate on Kubernetes resources exclusively - which presents a problem when trying to deploy resources defined in Bicep manifests. This design proposes the creation of a new Kubernetes controller in Radius called the DeploymentTemplate Controller that will allow users to deploy resources defined in Bicep manifests using Kubernetes tooling using Radius.
+
+## Today's Status
+
+Today, Radius users are unable to deploy resources defined in Bicep manifests using Kubernetes tooling. This is because Kubernetes tools do not understand Bicep manifests. Users must use the Radius CLI to deploy resources defined in Bicep manifests.
+
+## Terms and definitions
+
+**CRD (Custom Resource Definition)**: A definition for a Kubernetes resource that allows users to define their own resource types.
+
+**CR (Custom Resource)**: An instance of a CRD that represents a custom resource in Kubernetes.
+
+**Kubernetes Tools**: Tools that exist in the Kubernetes ecosystem and operate on kubernetes resources. Examples include `helm`, GitOps tools such as `Flux` and `ArgoCD`, and `kubectl`.
+
+**Bicep**: An infrastructure-as-code language that when used with Radius, can deploy Radius resources, Azure resources, and AWS resources.
+ 
+## Objectives
+
+### Goals
+
+**Goal: Users can use Kubernetes tooling to continuously deploy and manage resources defined in Bicep manifests**
+- With this work, users will be able to deploy resources defined in Bicep using only Kubernetes. We will essentially be providing a "translation layer" between Kubernetes resources (that Kubernetes tools can understand) and Radius resources (that Radius can understand).
+- With this work, users can rely on the `DeploymentTemplate` controller to repair drift from the desired state, and to perform disaster recovery. These are benefits users can't get from `rad deploy` or `az deployment create`.
+
+**Goal: Users can quickly generate a Kubernetes Custom Resource from Bicep using the Radius CLI**
+- We will provide a CLI command that generates the DeploymentTemplate resource from a Bicep manifest to make this feature easy to adopt.
+
+**Goal: The behavior of DeploymentTemplate is consistent with user expectations for a Kubernetes-enabled tool**
+- We will follow user expectations for Kubernetes controllers and resources when designing the DeploymentTemplate controller and resource. This includes building in support for retries, status updates, and error handling.
+
+### Non-goals
+
+**Non-goal (out of scope): Full support for GitOps**
+
+- We will not yet be implementing automatic generation of DeploymentTemplate resources from Bicep manifests or querying Git repositories. This design will enable this work, and it will be covered in a future design document.
```

Comment:
Do you mean understanding if there is an update to a Bicep file in a repository and automatically creating the DeploymentTemplate for that update and deploy it?